reportportal_client/core/worker.py

```diff
@@ -0,0 +1,169 @@
+
+    def _command_process(self, cmd):
+        """Process control command sent to the worker.
+
+        :param cmd: ControlCommand to be processed
+        """
+        if not cmd:
+            return  # No command received
+
+        logger.debug('[%s] Processing {%s} command', self.name, cmd)
+        if cmd == ControlCommand.REPORT_STATUS:
+            logger.debug('[%s] Current status for tasks is: {%s} unfinished',
+                         self.name, self._data_queue.unfinished_tasks)
+
+        if cmd == ControlCommand.STOP:
+            request = self._request_get()
+            while request is not None:
+                self._request_process(request)
+                request = self._request_get()
+
+        if cmd and cmd.is_stop_cmd():
+            self._stop()
+
+    def _monitor(self):
+        """Monitor worker queues and process them.
+
+        This method runs on a separate, internal thread. The thread will
+        terminate if the stop_immediate control command is received. If
+        the stop control command is sent, the worker will process all the
+        items from the queue before terminate.
+        """
+        while True:
+            cmd = self._command_get()
+            self._command_process(cmd)
+
+            if cmd and cmd.is_stop_cmd():
+                logger.debug('[%s] Exiting due to {%s} command',
+                             self.name, cmd)
+                break
+
+            request = self._request_get()
+            self._request_process(request)
+
+    def _request_get(self):
+        """Get response object from the data queue."""
+        try:
+            _, request = self._data_queue.get_nowait()
+            logger.debug('[%s] Received {%s} request', self.name, request)
+            return request
+        except Empty:
+            return None
+
+    def _request_process(self, request):
+        """Send request to RP and update response attribute of the request."""
+        if not request:
+            return  # No request received
+
+        logger.debug('[%s] Processing {%s} request', self.name, request)
+        request.response = request.http_request.make()
+        self._data_queue.task_done()
+
+    def _stop(self):
+        """Routine that stops the worker thread(s).
+
+        This asks the thread to terminate, and then waits for it to do so.
+        Note that if you don't call this before your application exits, there
+        may be some records still left on the queue, which won't be processed.
+        """
+        if self._thread.isAlive() and self._thread is not currentThread():
+            self._thread.join()
+        self._thread = None
+
+    def send_command(self, cmd):
+        """Send control command to the worker queue."""
+        self._cmd_queue.put(cmd)
+
+    def send_request(self, request):
+        """Send a request to the worker queue.
+
+        :param request: RPRequest object
+        """
+        self._data_queue.put((request.priority, request))
```

Comment:
We won't need to separate priority if we implement a priority field comparator for the request (__lt__)
```
from queue import PriorityQueue


class Request:
    def __init__(self, weight=1):
        self.weight = weight

    def __lt__(self, other):
        return self.weight < other.weight

    def __str__(self):
        return f"Request<{self.weight}>"


q = PriorityQueue()

q.put(Request(2))
q.put(Request(3))
q.put(Request(4))
q.put(Request(1))
q.put(Request(4))
q.put(Request(5))

while True:
    print(q.get())
```

>>>
```
Request<1>
Request<2>
Request<3>
Request<4>
Request<4>
Request<5>
```